lib/loglevel.js

```diff
@@ -189,68 +195,108 @@
       self.methodFactory = factory || defaultMethodFactory;
 
       self.getLevel = function () {
-          if (userLevel != null) {
-            return userLevel;
-          } else if (defaultLevel != null) {
-            return defaultLevel;
-          } else {
-            return inheritedLevel;
+          if (userLevel !== null) {
+              return userLevel;
+          }
+          if (defaultLevel !== null) {
+              return defaultLevel;
           }
+          if (inheritedLevel === null) {
+              inheritedLevel = self.getInheritedLevel();
+          }
+          return inheritedLevel;
       };
 
-      self.setLevel = function (level, persist) {
-          userLevel = normalizeLevel(level);
-          if (persist !== false) {  // defaults to true
-              persistLevelIfPossible(userLevel);
+      self.getInheritedLevel = function() {
+          if( defaultLogger && self.parentKey ) {
+              return defaultLogger.getLogger(self.parentKey).getLevel();
           }
-
-          // NOTE: in v2, this should call rebuild(), which updates children.
-          return replaceLoggingMethods.call(self);
+          if( !defaultLogger || defaultLogger===self ) {
+              return normalizeLevel("WARN");
+          }
+          return normalizeLevel(
+              defaultLogger ? defaultLogger.getLevel() : "WARN" );
       };
 
-      self.setDefaultLevel = function (level) {
-          defaultLevel = normalizeLevel(level);
-          if (!getPersistedLevel()) {
-              self.setLevel(level, false);
+      /** The parentKey is the name of the parent this inherits from */
+      self.parentKey = undefined;
```

Comment:
Loggers here store separate variables for `name`, `categoriesParam`, `parentKey` and `categories` as a separate public field, and `newCategories` & `childCategories` fields added in getLogger. All of those are fundamentally derived data of a single thing.

I think we only need one of these: Logger should take a single `keyHierarchy` that's an `Array<string | Symbol>` (no separate name + hierarchy).

That is effectively `categoriesParam`. `parentKey` is then `keyHierarchy.slice(0, -1)`, name is `keyHierarchy.join('.')`. `categories`, `newCategories` and `childCategories` can go.

I think there's a general design simplification that will follow if we head along these lines - I'll try to expand on that in the comments below.